plugins/camera/camera_impl.cpp

```diff
@@ -0,0 +1,739 @@
+        _status.callback = nullptr;
+    }
+
+    _parent.refresh_timeout_handler(_status.timeout_cookie);
+}
+
+void CameraImpl::capture_info_async(Camera::capture_info_callback_t callback)
+{
+    std::lock_guard<std::mutex> lock(_capture_info.mutex);
+
+    _capture_info.callback = callback;
+}
+
+void CameraImpl::process_camera_capture_status(const mavlink_message_t &message)
+{
+    mavlink_camera_capture_status_t camera_capture_status;
+    mavlink_msg_camera_capture_status_decode(&message, &camera_capture_status);
+
+    {
+        std::lock_guard<std::mutex> lock(_status.mutex);
+
+        _status.data.video_on = (camera_capture_status.video_status == 1);
+        _status.data.photo_interval_on = (camera_capture_status.image_status == 2 ||
+                                          camera_capture_status.image_status == 3);
+        _status.received_camera_capture_status = true;
+    }
+
+    check_status();
+}
+
+void CameraImpl::process_storage_information(const mavlink_message_t &message)
+{
+    mavlink_storage_information_t storage_information;
+    mavlink_msg_storage_information_decode(&message, &storage_information);
+
+    {
+        std::lock_guard<std::mutex> lock(_status.mutex);
+        if (storage_information.status == 0) {
+            _status.data.storage_status = Camera::Status::StorageStatus::NOT_AVAILABLE;
+        } else if (storage_information.status == 1) {
+            _status.data.storage_status = Camera::Status::StorageStatus::UNFORMATTED;
+        } else if (storage_information.status == 2) {
+            _status.data.storage_status = Camera::Status::StorageStatus::FORMATTED;
+        }
+        _status.data.available_storage_mib = storage_information.available_capacity;
+        _status.data.used_storage_mib = storage_information.used_capacity;
+        _status.data.total_storage_mib = storage_information.total_capacity;
+        _status.received_storage_information = true;
+    }
+
+    check_status();
+}
+
+void CameraImpl::process_camera_image_captured(const mavlink_message_t &message)
+{
+    mavlink_camera_image_captured_t image_captured;
+    mavlink_msg_camera_image_captured_decode(&message, &image_captured);
+
+    {
+        std::lock_guard<std::mutex> lock(_capture_info.mutex);
+
+        if (_capture_info.callback) {
+            Camera::CaptureInfo capture_info = {};
+            capture_info.position.latitude_deg = image_captured.lat / 1e7;
+            capture_info.position.longitude_deg = image_captured.lon / 1e7;
+            capture_info.position.absolute_altitude_m = image_captured.alt / 1e3f;
+            capture_info.position.relative_altitude_m = image_captured.relative_alt / 1e3f;
+            capture_info.time_utc_us = image_captured.time_utc;
+            capture_info.quaternion.w = image_captured.q[0];
+            capture_info.quaternion.x = image_captured.q[1];
+            capture_info.quaternion.y = image_captured.q[2];
+            capture_info.quaternion.z = image_captured.q[3];
```

Comment:
How about having friend method:
```cpp
friend const Camera::CaptureInfo &operator=(Camera::CaptureInfo& to, float quaternion[4]);
```
so that we can do:
```cpp
capture_info.quaternion = image_captured.q;
```
I mean, may be later.

Reply:
Nice, didn't know this.